File: robocode2-bot-api-java/src/main/java/net/robocode2/mapper/BulletStateMapper.java
package net.robocode2.mapper;

import lombok.NonNull;
import lombok.experimental.UtilityClass;
import lombok.val;
import net.robocode2.BulletState;

import java.util.ArrayList;
import java.util.List;

/** Utility class for mapping a bot state */
@UtilityClass
class BulletStateMapper {

  BulletState map(@NonNull final net.robocode2.schema.BulletState source) {
    return BulletState.builder()
        .bulletId(source.getBulletId())
        .ownerId(source.getOwnerId())
        .power(source.getPower())
        .x(source.getX())
        .y(source.getY())
        .direction(source.getDirection())
        .speed(source.getSpeed())
        .build();
  }

  List<BulletState> map(@NonNull final List<net.robocode2.schema.BulletState> source) {
    val bulletStates = new ArrayList<BulletState>();
    source.forEach(bulletState -> bulletStates.add(map(bulletState)));
    return bulletStates;
  }
}

File: bot-api/python/src/robocode_tank_royale/bot_api/mapper/bullet_state_mapper.py
from typing import Sequence, overload

from ..bullet_state import BulletState
from robocode_tank_royale.schema import BulletState as SchemaBulletState


def _map_bullet_state(
    source: SchemaBulletState | None,
) -> BulletState | None:
    """Map a schema BulletState to a bot-api BulletState."""
    if source is None:
        return None

    return BulletState(
        source.bullet_id,
        source.owner_id,
        source.power,
        source.x,
        source.y,
        source.direction,
        source.color  # type: ignore
    )


class BulletStateMapper:
    """Utility class for mapping bullet states."""

    @overload
    @staticmethod
    def map(source: SchemaBulletState) -> BulletState:
        """Map a schema BulletState to a bot-api BulletState."""
        ...

    @overload
    @staticmethod
    def map(source: Sequence[SchemaBulletState | None]) -> Sequence[BulletState | None]:
        """Map a sequence of schema BulletState to a sequence of bot-api BulletState."""
        ...

    @overload
    @staticmethod
    def map(source: None) -> None:
        """Map a sequence of schema BulletState to a sequence of bot-api BulletState."""
        ...

    @staticmethod
    def map(
        source: SchemaBulletState | Sequence[SchemaBulletState | None] | None,
    ) -> BulletState | Sequence[BulletState | None] | None:
        """Map a schema BulletState to a bot-api BulletState."""
        if source is None:
            return None

        if isinstance(source, Sequence):
            return [_map_bullet_state(s) for s in source]
        return _map_bullet_state(source)
